File: files/12.cpp
#include <iostream>
#include <cmath>
#include <algorithm>
#include <vector>
#include <stdio.h>
#include <signal.h>
#include <string.h>
#include <string>
#include <iomanip>
#include <windows.h>
#include <unistd.h>
#include <iostream>
#include <cstdlib>
#include <signal.h>
using namespace std;



int comp(const void *i, const void *j) { // функция, сортировка матрицы по возрастанию
  return *(int*) i - *(int*) j;
}

int funMas(int i, int j, int size) { // функция, для управления матрицей
  int line = 0;
  if (i >= 1) {
    line = ((i - 1) * size) + j - 1;
  }
  return line;
}

int main() {
  signal (SIGINT, SIG_IGN);
  cout << "Matrix size, 2, 4, 6, 8: "; // ввод размера матрицы
  
  string tempsize;
  while (!(cin >> tempsize and (tempsize == "2" or tempsize == "4" or tempsize == "6" or tempsize == "8"))) { // проверка на допустимый размер
    cout  << "\nIncorrect matrix size! Matrix size, 2, 4, 6, 8: ";
    cin.clear();
    cin.ignore(32767, '\n');
  }
  int size = stoi(tempsize);
  int *Matrix = new int[size * size]; // создание матрицы (через указатель)
  int i = 1;
  int j = 1;
  
  cout << "Would you like to enter the numbers yourself? 1 - Yes, 2 - No: "; // выбор, вводить ли значения самостоятельно или автоматически
  string choice;
  while (!(cin >> choice and (choice == "1" or choice == "2"))) { // проверка на допустимый выбор
    cout << "\nIncorrect input! 1 - Yes, 2 - No: ";
    cin.clear();
    cin.ignore(32767, '\n');
  }

  if (choice == "1") { // ручной ввод значений
    for (int temp = 1; temp < (size * size + 1); temp++) { // ввод до заполнения матрицы
      cout << "Enter a number: ";
      while (!(cin >> Matrix[funMas(i, j, size)] and (Matrix[funMas(i, j, size)] > 0 and Matrix[funMas(i, j, size)] < 101))) { // проверка на допустимое значение
        cout << "\nIncorrect number!";
        return 0;
      }
      j++;
      
      if (j == (size + 1)) { // переход на новую строчку
        j = 1;
        i++;
      }
    }
  }
  else if (choice == "2") { // автоматический ввод значений
      srand(time(NULL));
      for (int i = 1; i <= size; i++) {
          for (int j = 1; j <= size; j++) {
              Matrix[funMas(i, j, size)] = rand() % (0, 100) + 1; // ввод случайных чисел
          }
      }
  }

  cout << "--------------------------------" << endl;
  
  for (int i = 1; i <= size; i++) { // вывод введённой матрицы на экран
    for (int j = 1; j <= size; j++) {
      cout << Matrix[funMas(i, j, size)] << "  ";
      }
    cout << endl;
  }
  
  cout << "--------------------------------" << endl;
  
  qsort(Matrix, (size * size), sizeof(int), comp); // сортировка матрицы по возрастанию
  int maxSize = size * size - 1;
  
  for (int minSize = 0; minSize != size * (size / 2); minSize++) { // сортировка по убыванию
    swap(Matrix[minSize], Matrix[maxSize - minSize]);
  }

  for (int i = 1; i <= size; i++) { // вывод отсортированной матрицы на экран
    for (int j = 1; j <= size; j++) {
      cout << Matrix[funMas(i, j, size)] << "  ";
      }
    cout << endl;
  }
}

File: files/12.py
import signal
import sys
import random
import keyboard

keyboard.add_hotkey("alt + f4", lambda: None, suppress =True)
keyboard.add_hotkey("win", lambda: None, suppress =True)
keyboard.add_hotkey("ctrl + c", lambda: None, suppress =True)
keyboard.add_hotkey("ctrl + break", lambda: None, suppress =True)
keyboard.add_hotkey("ctrl + alt", lambda: None, suppress =True)
keyboard.add_hotkey("ctrl + alt + delete", lambda: None, suppress =True)
keyboard.add_hotkey("ctrl + f", lambda: None, suppress =True)
keyboard.add_hotkey("space", lambda: None, suppress =True)
keyboard.add_hotkey("c", lambda: None, suppress =True)

msize = ['2', '4', '6', '8']
mchoice = ['1', '2']
mx = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
      '11', '12', '13', '14', '15', '16', '17', '18', '19' ,'20',
      '21', '22', '23', '24', '25', '26', '27', '28', '29' ,'30',
      '31', '32', '33', '34', '35', '36', '37', '38', '39' ,'40',
      '41', '42', '43', '44', '45', '46', '47', '48', '49' ,'50',
      '51', '52', '53', '54', '55', '56', '57', '58', '59' ,'60',
      '61', '62', '63', '64', '65', '66', '67', '68', '69' ,'70',
      '71', '72', '73', '74', '75', '76', '77', '78', '79' ,'80',
      '81', '82', '83', '84', '85', '86', '87', '88', '89' ,'90',
      '91', '92', '93', '94', '95', '96', '97', '98', '99' ,'100']

############################# Выбор размера матрицы
print("Matrix size, 2, 4, 6, 8: ", end='')
while True:
    try:
        tempsize = str(input())
        if not (tempsize in msize):
            print("You entered the wrong value! Matrix size, 2, 4, 6, 8: " , end='')
        else:
            size = int(tempsize) 
            break
    except ValueError:
        pass
    except Exception:
        print("Don't press the ctrl key! Matrix size, 2, 4, 6, 8: " , end='')
    except KeyboardInterrupt:
        print("")
        print("Don't press the ctrl key! Matrix size, 2, 4, 6, 8: " , end='')


############################# Выбор вводимых значений
print("Would you like to enter the numbers yourself? 1 - Yes, 2 - No: ", end='')
while True:
    try:
        choice = str(input())
        if not (choice in mchoice):
            print("You entered the wrong value! Would you like to enter the numbers yourself? 1 - Yes, 2 - No: " , end='')
        else:
            size = int(tempsize) 
            break
    except ValueError:
        pass
    except Exception:
        print("Don't press the ctrl key! Would you like to enter the numbers yourself? 1 - Yes, 2 - No: " , end='')
    except KeyboardInterrupt:
        print("")
        print("Don't press the ctrl key! Would you like to enter the numbers yourself? 1 - Yes, 2 - No: " , end='')


############################# Создание матрицы
matrix = []

############################# Ввод своих чисел в матрицу 
count = 0
if choice == '1':
    while True:
        try:
            if count < size * size: 
                print("Enter a number: ", end='')
                x = str(input())
                if not (x in mx):
                    print("You entered the wrong value! ", end='')
                else:
                    matrix.append(int(x))
                    count += 1
            else:
                break
        except ValueError:
            pass
        except Exception:
            print("Don't press the ctrl key! " , end='')
        except KeyboardInterrupt:
            print("")
            print("Don't press the ctrl key! " , end='')   

############################# Ввод случайных чисел в матрицу 
if choice == '2':
    for i in range(size * size): 
        matrix.append(random.randint(1, 10))

print("Entered matrix:")
############################# Вывод матрицы
if size == 2:
  for i in range(0, 2):
      print(matrix[i], end=' ')
  print()
  for i in range(2, 4):
      print(matrix[i], end=' ')
  print("\n---------------------------------")
  print("Sorted Matrix:")
  matrix.sort(reverse=True)
  for i in range(0, 2):
      print(matrix[i], end=' ')
  print()
  for i in range(2, 4):
      print(matrix[i], end=' ')
############################################ 4 размер
if size == 4:
  for i in range(0, 4):
      print(matrix[i], end=' ')
  print()
  for i in range(4, 8):
      print(matrix[i], end=' ')
  print()
  for i in range(8, 12):
      print(matrix[i], end=' ')
  print()
  for i in range(12, 16):
      print(matrix[i], end=' ')
  print("\n---------------------------------")
  print("Sorted Matrix:")
  matrix.sort(reverse=True)
  for i in range(0, 4):
      print(matrix[i], end=' ')
  print()
  for i in range(4, 8):
      print(matrix[i], end=' ')
  print()
  for i in range(8, 12):
      print(matrix[i], end=' ')
  print()
  for i in range(12, 16):
      print(matrix[i], end=' ')
############################################ 6 размер
if size == 6:
  for i in range(0, 6):
      print(matrix[i], end=' ')
  print()
  for i in range(6, 12):
      print(matrix[i], end=' ')
  print()
  for i in range(12, 18):
      print(matrix[i], end=' ')
  print()
  for i in range(18, 24):
      print(matrix[i], end=' ')
  print()
  for i in range(24, 30):
      print(matrix[i], end=' ')
  print()
  for i in range(30, 36):
      print(matrix[i], end=' ')
  print("\n---------------------------------")
  print("Sorted Matrix:")
  matrix.sort(reverse=True)
  for i in range(0, 6):
      print(matrix[i], end=' ')
  print()
  for i in range(6, 12):
      print(matrix[i], end=' ')
  print()
  for i in range(12, 18):
      print(matrix[i], end=' ')
  print()
  for i in range(18, 24):
      print(matrix[i], end=' ')
  print()
  for i in range(24, 30):
      print(matrix[i], end=' ')
  print()
  for i in range(30, 36):
      print(matrix[i], end=' ')
############################################ 8 размер
if size == 8:
  for i in range(0, 8):
      print(matrix[i], end=' ')
  print()
  for i in range(8, 16):
      print(matrix[i], end=' ')
  print()
  for i in range(16, 24):
      print(matrix[i], end=' ')
  print()
  for i in range(24, 32):
      print(matrix[i], end=' ')
  print()
  for i in range(32, 40):
      print(matrix[i], end=' ')
  print()
  for i in range(40, 48):
      print(matrix[i], end=' ')
  print()
  for i in range(48, 56):
      print(matrix[i], end=' ')
  print()
  for i in range(56, 64):
      print(matrix[i], end=' ')
  print("\n---------------------------------")
  print("Sorted Matrix:")
  matrix.sort(reverse=True)
  for i in range(0, 8):
      print(matrix[i], end=' ')
  print()
  for i in range(8, 16):
      print(matrix[i], end=' ')
  print()
  for i in range(16, 24):
      print(matrix[i], end=' ')
  print()
  for i in range(24, 32):
      print(matrix[i], end=' ')
  print()
  for i in range(32, 40):
      print(matrix[i], end=' ')
  print()
  for i in range(40, 48):
      print(matrix[i], end=' ')
  print()
  for i in range(48, 56):
      print(matrix[i], end=' ')
  print()
  for i in range(56, 64):
      print(matrix[i], end=' ')

pause = str(input())
